File: shared/types/events.ts
export type Severity = "info" | "warning" | "critical";

export interface SystemEvent {
  id: string;
  type: string;
  timestamp: number;
  source: string;
  severity: Severity;
  payload: any;
}

export interface TraceEvent extends SystemEvent {
  type: "trace:new";
  payload: {
    category: 'EXECUTION' | 'RETRIEVAL' | 'AUDITS' | 'DRIFT' | 'FAILURES';
    message: string;
    traceId: string;
  };
}

export interface WorkflowEvent extends SystemEvent {
  type: "workflow:update";
  payload: {
    workflowId: string;
    name: string;
    status: 'IDLE' | 'RUNNING' | 'COMPLETED' | 'FAILED';
    progress: number;
    step?: string;
  };
}

export interface MetricEvent extends SystemEvent {
  type: "metric:update";
  payload: {
    promptOverhead: number;
    memoryInjection: number;
    retrievalDuplication: number;
    agentChatter: number;
    semanticRedundancy: number;
  };
}

File: baton_server/schemas/events.py
from pydantic import BaseModel
from typing import Any, Optional, Literal
from datetime import datetime
import uuid

Severity = Literal["info", "warning", "critical"]


class SystemEvent(BaseModel):
    id: str = str(uuid.uuid4())
    type: str
    timestamp: float = datetime.now().timestamp()
    source: str
    severity: Severity = "info"
    payload: Any


class TracePayload(BaseModel):
    category: Literal["EXECUTION", "RETRIEVAL", "AUDITS", "DRIFT", "FAILURES"]
    message: str
    traceId: str


class WorkflowPayload(BaseModel):
    workflowId: str
    name: str
    status: Literal["IDLE", "RUNNING", "COMPLETED", "FAILED"]
    progress: float
    step: Optional[str] = None


class MetricPayload(BaseModel):
    promptOverhead: float
    memoryInjection: float
    retrievalDuplication: float
    agentChatter: float
    semanticRedundancy: float
